### ISSUE TITLE
Help for SHARED_LIBRARIES: libqdMetaData and libqservice

### ISSUE THREAD

**Issue by CoolestEnoch** [datenpanne/android_device_huawei_t2-common#1]
TL;DR
I am from a non-English-spoken country so sorry for these grammar mistakes. Though my college had German classes but I didn't study well then... so I read threads from Android-Hilfe forum always need a translator XD.... I would be grateful if you could reply me here in English. XD
My device is Huawei FDR-A01w (Federer), which successfully installed LineageOS 16, 17 and 18 from android-hilfe forum, now running on [LineageOS 18 (hilfe)](https://www.android-hilfe.de/forum/root-custom-roms-modding-fuer-huawei-mediapad-t2-10-0-pro.3369/lineageos-18-1-unofficial.999873-page-2.html). My device's stock ROM is FDR-A01wC635B001D001 (iFlytek customised ROM), bootloader is unlocked.
 I am interested in this tablet because it carries my three years of memories and accompanies me through these three years. It is greatly appreciated that you have made these ROMs so that it can have the best compatibility with latest softwares.

Environment:
Computer: Intel core i5-10300H, 15.4 GB RAM and 16 GB swap, Samsung Evo 870 1TB SSD, Debian 12, zram is enabled.</br>
~/android/lineage/.repo/local_manifest/federer.xml:
``` xml
<?xml version="1.0" encoding="UTF-8"?>
<manifest>
    <remote name="vendor" fetch="https://github.com" revision="la-17.1" />
    <remote name="device" fetch="https://github.com" revision="la-17.1"/>
    <remote name="device-common" fetch="https://github.com" revision="la-17.1"/>
    <remote name="kernel" fetch="https://github.com" revision="LaOS-15.1"/>
    <remote name="sepolicy_legacy" fetch="https://github.com" revision="lineage-20.0"/>
    <!-- Vendor folders -->
    <project path="vendor/huawei/federer" name="datenpanne/android_vendor_huawei_federer" remote="vendor" />
    <!--Device Trees-->
    <project path="device/huawei/federer" name="datenpanne/android_device_huawei_federer" remote="device" />
    <project path="device/huawei/t2-common" name="datenpanne/android_device_huawei_t2-common" remote="device-common" />
    <!--Kernel-->
    <project path="kernel/huawei/federer" name="datenpanne/android_kernel_huawei_federer" remote="kernel" />
    <!--SePolicy Legacy-->
    <project path="device/qcom/sepolicy-legacy" name="LineageOS/android_device_qcom_sepolicy" remote="sepolicy_legacy" />
</manifest>
```

Problem:
I am using this repository try to build and compile LineageOS 20, and I encountered a problem. When I tried to compile at first, it says
``` text
device/huawei/t2-common/camera/QCamera2/HAL/Android.mk: error: "camera.msm8916 (SHARED_LIBRARIES android-arm) missing libqdMetaData (SHARED_LIBRARIES android-arm)"
You can set ALLOW_MISSING_DEPENDENCIES=true in your environment if this is intentional, but that may defer real problems until later in the build.
device/huawei/t2-common/camera/QCamera2/HAL/Android.mk: error: "camera.msm8916 (SHARED_LIBRARIES android-arm) missing libqservice (SHARED_LIBRARIES android-arm)"
You can set ALLOW_MISSING_DEPENDENCIES=true in your environment if this is intentional, but that may defer real problems until later in the build.
```
And then I set `ALLOW_MISSING_DEPENDENCIES=true` in my environment and re-compile with `brunch federer` then.

After some minutes, it encountered a new problem, which says:
``` text
FAILED: ninja: 'out/target/product/federer/obj_arm/SHARED_LIBRARIES/libqdMetaData_intermediates/libqdMetaData.so.toc', needed by 'out/target/product/federer/obj_arm/SHARED_LIBRARIES/camera.msm8916_intermediates/LINKED/camera.msm8916.so', missing and no known rule to make it
```
I searched `libqdmetadata` in my directory `~/android/lineage` and I got these files:
``` bash
┌─[android233@handsome] - [~/android/lineage] - [一 1月 22, 22:57]
└─[$] <> find . -name "*libqdmetadata*"   
./vendor/qcom/opensource/commonsys-intf/display/libqdmetadata
./vendor/qcom/opensource/display/libqdmetadata
./vendor/qcom/opensource/commonsys/display/libqdmetadata
./hardware/qcom-caf/sm8550/display/libqdmetadata
./hardware/qcom-caf/sm8450/display/libqdmetadata
```

I suddenly found that the new issue was caused by the environment parameter which I set before. So I tried to see the source code of `~/android/lineage/device/huawei/t2-common/camera/QCamera2/HAL/Android.mk` and found at line 69, which says `LOCAL_SHARED_LIBRARIES += libqdMetaData libqservice libbinder` there. And in this file, from line 35-48 says
``` makefile
LOCAL_C_INCLUDES := \
	$(LOCAL_PATH)/../stack/common \
	$(LOCAL_PATH)/../../mm-image-codec/qexif \
	$(LOCAL_PATH)/../../mm-image-codec/qomx_core \
	$(LOCAL_PATH)/../util \
	$(LOCAL_PATH)/wrapper \
        $(LOCAL_PATH)/../stack/mm-camera-interface/inc \
        $(TARGET_OUT_HEADERS)/mm-camera-lib/cp/prebuilt \
        hardware/qcom-caf/msm8916/display/libgralloc \
        hardware/qcom-caf/msm8916/display/libqdutils \
        hardware/qcom-caf/msm8916/display/libqservice \
        hardware/qcom-caf/msm8916/media/libstagefrighthw \
        hardware/qcom-caf/msm8916/media/mm-core \
        system/media/camera/include
```
But I tried to reach the directory `~/android/lineage/hardware/qcom-caf/msm8916` but failed, the last folder `msm8916` isn't there. So why it only throw an error about `libqdMetaData` and `libqservice`? How do this MakeFile find and use these include path from a doesn't exist directory? I found that this makefile is always the same in all branches here...

I am doubting about which shared libraries it uses (I mean, the relative path to ~/android/lineage), and can I use these search-results above, and how to use or configure that....I'm stuck here...

And... can you help me with these issues..? Or tell me something about how to fix this dependencies missing..? I'll try my best to provide the logcats that is required.Thanks a lot!! 
XD

**Comment by datenpanne** [comment]
Hello, sorry my late answer.

As I know our kernel isn't supported anymore, because is to old, last supported version is LA18
OK, this issue has nothing to do with camera, but bevor you invest much time in this, o just telling you.

You can read it here https://www.xda-developers.com/lineageos-19-android-12/

In section "eBPF over iptables".

Next time PM me, because this repository is working and has nothing to do with LA20.

**Comment by datenpanne** [comment]
Wrong repository.

**Comment by CoolestEnoch** [comment]
Oh thank you for your reply 😂 I got it
